File: android/src/main/java/sa/vrtx/reactnative/VrtxAndroidModule.kt
package sa.vrtx.reactnative

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.app.ContextHolder
import sa.vrtx.public.Vrtx
import sa.vrtx.public.configuration.Environment
import sa.vrtx.public.configuration.Language
import sa.vrtx.public.configuration.ThemeMode
import androidx.compose.ui.text.font.FontFamily
import androidx.activity.ComponentActivity

class VrtxAndroidModule : Module() {
  
  private fun getActivity(): ComponentActivity? {
    return ContextHolder.getActivityProvider()?.currentActivity as? ComponentActivity
  }
  
  override fun definition() = ModuleDefinition {
    Name("VrtxAndroid")

    Constant("LIBRARY_NAME") {
      "vrtx-android"
    }

    Events("onSuccess", "onError")

    AsyncFunction("setup") { 
      clientId: String, 
      clientSecret: String, 
      environment: String, 
      language: String, 
      themeMode: String?,
      promise: Promise ->
      
      val env = when(environment.uppercase()) {
        "STAGING" -> Environment.Staging
        else -> Environment.Sandbox
      }
      
      val lang = when(language.uppercase()) {
        "ARABIC" -> Language.Arabic
        else -> Language.English
      }
      
      val theme = when(themeMode?.uppercase()) {
        "DARK" -> ThemeMode.DARK
        else -> ThemeMode.LIGHT
      }
      
      val fontFamily = FontFamily.Default
      
      // Get activity and run on UI thread
      val activity = getActivity()
      if (activity == null) {
        promise.reject(CodedException("VRX_ERROR", "Activity not available", null))
        return@AsyncFunction
      }
      
      activity.runOnUiThread {
        try {
          Vrtx.setup(
            clientId = clientId,
            clientSecret = clientSecret,
            environment = env,
            language = lang,
            themeMode = theme,
            fontFamily = fontFamily,
            onSuccess = {
              promise.resolve(null)
              sendEvent("onSuccess", null)
            },
            onError = { error ->
              val errorMessage = error.message ?: "Unknown error"
              promise.reject(CodedException("VRX_ERROR", errorMessage, null))
              sendEvent("onError", mapOf("code" to "VRX_ERROR", "message" to errorMessage))
            }
          )
        } catch (e: Exception) {
          val errorMessage = e.message ?: "Failed to initialize Vrtx SDK"
          promise.reject(CodedException("VRX_ERROR", errorMessage, e))
          sendEvent("onError", mapOf("code" to "VRX_ERROR", "message" to errorMessage))
        }
      }
    }
  }
}

File: ios/VrtxAndroidModule.swift
import ExpoModulesCore
import VRTX

public class VrtxAndroidModule: Module {
  public func definition() -> ModuleDefinition {
    Name("VrtxAndroid")

    Constant("LIBRARY_NAME") {
      return "vrtx-ios"
    }

    Events("onSuccess", "onError")

    /**
     * Initialize and launch the Vrtx SDK UI flow on iOS via the vrtx-ios
     * SDK (VRTX.xcframework). Mirrors the Android Vrtx.setup contract.
     */
    AsyncFunction("setup") { (
      clientId: String,
      clientSecret: String,
      environment: String,
      language: String,
      mode: String?,
      fontFamily: String?,
      promise: Promise
    ) in
      let env: Environment = (environment.uppercased() == "STAGING") ? .staging : .sandbox
      let lang: Language = (language.uppercased() == "ARABIC") ? .arabic : .english
      let theme: Mode = (mode?.uppercased() == "DARK") ? .dark : .light
      let font = fontFamily ?? ""

      DispatchQueue.main.async { [weak self] in
        guard let self else { return }
        Vrtx.setup(
          environment: env,
          clientID: clientId,
          clientSecret: clientSecret,
          mode: theme,
          language: lang,
          fontFamily: font,
          onSuccess: {
            promise.resolve(nil)
            self.sendEvent("onSuccess")
          },
          onError: { error in
            let message = error.message
            promise.reject("VRX_ERROR", message)
            self.sendEvent("onError", [
              "code": "VRX_ERROR",
              "message": message
            ])
          }
        )
      }
    }
  }
}
